pkg/controller/workloadspread/reschedule.go

```diff
@@ -0,0 +1,137 @@
+// sufficient resource can lead to some Pods scheduled failed. WorkloadSpread has multiple subset, so these
+// unschedulable Pods should be rescheduled to other subsets.
+// Controller will mark the subset contains unschedulable Pods to unschedulable status and Webhook cannot inject
+// Pod into this subset by check subset's status. The unschedulable subset' status can be kept for 5 minutes and then
+// should be recovered schedulable status to try scheduling Pods again.
+// TODO optimize the unschedulable duration of subset.
+// return one parameters - unschedulable Pods belongs to this subset.
+func (r *ReconcileWorkloadSpread) rescheduleSubset(ws *appsv1alpha1.WorkloadSpread,
+	pods []*corev1.Pod,
+	subsetStatus, oldSubsetStatus *appsv1alpha1.WorkloadSpreadSubsetStatus) []*corev1.Pod {
+	scheduleFailedPods := make([]*corev1.Pod, 0)
+	for i := range pods {
+		if PodUnscheduledTimeout(ws, pods[i]) {
+			scheduleFailedPods = append(scheduleFailedPods, pods[i])
+		}
+	}
+	unschedulable := len(scheduleFailedPods) > 0
+	if unschedulable {
+		klog.V(3).Infof("Subset (%s) of WorkloadSpread (%s/%s) is unschedulable", subsetStatus.Name, ws.Namespace, ws.Name)
+	}
+
+	oldCondition := GetWorkloadSpreadSubsetCondition(oldSubsetStatus, appsv1alpha1.SubsetSchedulable)
+	if oldCondition == nil {
+		if unschedulable {
+			SetWorkloadSpreadSubsetCondition(subsetStatus, NewWorkloadSpreadSubsetCondition(appsv1alpha1.SubsetSchedulable, corev1.ConditionFalse, "", ""))
+		} else {
+			SetWorkloadSpreadSubsetCondition(subsetStatus, NewWorkloadSpreadSubsetCondition(appsv1alpha1.SubsetSchedulable, corev1.ConditionTrue, "", ""))
+		}
+		return scheduleFailedPods
+	}
+	// initialize to old condition.
+	SetWorkloadSpreadSubsetCondition(subsetStatus, oldCondition.DeepCopy())
+
+	if unschedulable {
+		SetWorkloadSpreadSubsetCondition(subsetStatus, NewWorkloadSpreadSubsetCondition(appsv1alpha1.SubsetSchedulable, corev1.ConditionFalse, "", ""))
+	} else {
+		if oldCondition.Status == corev1.ConditionFalse {
+			expectReschedule := oldCondition.LastTransitionTime.Add(ScheduledFailedDuration)
+			currentTime := time.Now()
+			// the duration of unschedule status more than 5 minutes, recover to schedulable.
+			if expectReschedule.Before(currentTime) {
+				r.recorder.Eventf(ws, corev1.EventTypeNormal,
+					"RecoverSchedulable", "Subset %s of WorkloadSpread %s/%s is recovered from unschedulable to schedulable",
+					subsetStatus.Name, ws.Namespace, ws.Name)
+				SetWorkloadSpreadSubsetCondition(subsetStatus, NewWorkloadSpreadSubsetCondition(appsv1alpha1.SubsetSchedulable, corev1.ConditionTrue, "", ""))
+			} else {
+				// less 5 minutes, keep unschedulable.
+				durationStore.Push(getWorkloadSpreadKey(ws), expectReschedule.Sub(currentTime))
+			}
+		}
+	}
+
+	return scheduleFailedPods
+}
+
+func (r *ReconcileWorkloadSpread) cleanupUnscheduledPods(ws *appsv1alpha1.WorkloadSpread,
```

Comment:
why cleanupUnscheduledPods is not been called inside the reschedule method? It feels like the logic should be highly cohesive

Reply:
The unscheduable Pods should be removed after the workloadspread's status has been updated, otherwise the Pods will be scheduled to the original subset and become unscheduable again. Therefore, we clean up unscheduable Pods in the end.